staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -404,14 +413,38 @@ func (r *crdHandler) removeDeadStorage() {
 		}
 		if !found {
 			klog.V(4).Infof("Removing dead CRD storage for %s/%s", s.spec.Group, s.spec.Names.Kind)
-			for _, storage := range s.storages {
-				// destroy only the main storage. Those for the subresources share cacher and etcd clients.
-				storage.CustomResource.DestroyFunc()
-			}
+			oldInfos = append(oldInfos, s)
 			delete(storageMap2, uid)
 		}
 	}
 	r.customStorage.Store(storageMap2)
+
+	for _, s := range oldInfos {
+		go tearDown(s)
+	}
+}
+
+// Wait up to a minute for requests to drain, then tear down storage
+func tearDown(oldInfo *crdInfo) {
+	requestsDrained := make(chan struct{})
+	go func() {
+		defer close(requestsDrained)
+		// Allow time for in-flight requests with a handle to the old info to register themselves
+		time.Sleep(time.Second)
+		// Wait for in-flight requests to drain
+		oldInfo.waitGroup.Wait()
```

Comment:
Too bad we can't context cancel the in-flight requests.

Reply:
they'll get timed out at the configured request timeout by https://github.com/kubernetes/kubernetes/blob/0252a323423080a3a230ab3053fdaaca8a972eb1/staging/src/k8s.io/apiserver/pkg/server/config.go#L552

Reply:
ah, nice